### PR TITLE
Session refactor

### PR DESCRIPTION
**Pull request by rvillarreal**
This addresses EPICENTER-1763 and EPICENTER-1751. It also has a centralized session management that validates the current account and project to determine if the session is valid.

### CODE REVIEW

File: src/managers/auth-manager.js [hunk 1]
```diff
@@ -357,7 +320,39 @@ AuthManager.prototype = $.extend(AuthManager.prototype, {
...
+     *      authMgr.addGroup('acme', { groupName: });
+     *
+     * **Parameters**
+     * @param {Object} `group` (Required) must have the project, groupName and groupId properties, isFac is optional and defaults to false
+     */
+    addGroup: function (group) {
+        return this.addGroups([group]);
+    },
+
+    addGroups: function (groups) {
```
Comment (narenranjit) [inline src/managers/auth-manager.js]: Why not have a single function which works with array/non-array similar to other services (like for instance `addUsers` in world-api-adapter)?

File: src/util/option-utils.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    store: {}
+};
+
+var optionUtils = {
+    /**
+     * Gets the final options by overriding the global options set with
+     * optionUtils#setDefaults() and the lib defaults.
+     * @param {object} `options` The final options object.
+     */
+    getOptions: function (options) {
```
Comment (narenranjit) [inline src/util/option-utils.js]: Nitpick on the syntax - `getOptions(options)` seems a little black-magicy where I can't tell what I'm getting back. Maybe `getMergedOptions`?

**Comment by mmrj** [comment]
running on local.forio.com, i tried the following: 
```
var authMgr = new F.manager.AuthManager({ /* varies by test case */ });
authMgr.login().then( function() {
    var tmp = authMgr.getCurrentUserSessionInfo();
    console.log("user session info: ", tmp);
    authMgr.logout().then(function() { console.log("logged out"); });
  }).fail( function() { console.log("login failed"); });
```

*epicenter.js 1.7.1*, 
1. pass in account, userName, password of a team member: logs in, logs out
2. pass in account, userName, password of an end user in one group in this team: logs in, logs out
3. pass in account, userName, password of an end user in multiple groups in this team: logs in, logs out
4. pass in account, userName, password, project of an end user in multiple groups in this team/project: login fails, because we don't know which group

*epicenter-edge-pr-157*
1. team member login fails
2. end user in one group logs in, logs out
3. end user in multiple groups login fails
4. end user in multiple groups with project specified login fails

i think that (1) is definitely a problem, and i think that (3) might be -- i didn't think we were planning to change the behavior here, but i'm not certain.

additionally, in all cases i'm not seeing a cookie. that's the expected because i'm running locally and haven't set it, correct?

@rvillarreal will keep testing and let you know if i find anything else. cc @narenranjit

**Comment by mmrj** [comment]
talked w/ @rvillarreal, things to change:
1. epicenterjs.session path is currently defaulting just to "/" but should default to "/app/accountid/projectid/". (check epicenter-scenario path too, just in case -- not tested yet)
2. authMgr.requiresGroup flag, when true, should only take effect if project is specified.
3. authMgr.addGroup vs addGroups from Naren above -- make just one method for addGroups

**Comment by rvillarreal** [comment]
On point 2. if project is passed in either the `new AuthManager()` or `login()` but is in the URL then it will act as it the project was passed.

**Comment by mmrj** [comment]
@rvillarreal now i'm seeing problems with three areas:
 - `authMgr.addGroups({ project: 'hello_world_python', groupName: 'js-libs-team2' });` fails if I log in as an end user in multiple groups & projects without specifying group or project
 - separately, when running from a different server (epistage1.foriodev.com), logging in as end user in many groups (no group or projects specified) fails. some variations of this may be due to platform version but not sure. let's look either during the sprint demo mtg or right after.

**Comment by mmrj** [comment]
adding groups is now separate ticket (EPICENTER-1939), and discussed that logging in as end user is correctly disallowed when on actual server b/c project is in uri, and if there is a project, then you need a group specified if you're in many. see additional notes in EPICENTER-1763. 

merging.

**Pull request merged**